pkg/sbproxy/notifications/producer.go

```diff
@@ -0,0 +1,304 @@
+		return fmt.Errorf("ProducerSettings: min ping period must be positive duration")
+	}
+	if p.ReconnectDelay < 0 {
+		return fmt.Errorf("ProducerSettings: reconnect delay must be non-negative duration")
+	}
+	if p.PongTimeout <= 0 {
+		return fmt.Errorf("ProducerSettings: pong time must be positive duration")
+	}
+	if p.PingPeriodPercentage <= 0 || p.PingPeriodPercentage >= 100 {
+		return fmt.Errorf("ProducerSettings: ping period percentage must be between 0 and 100")
+	}
+	if p.NotificationsAPIPath == "" {
+		return fmt.Errorf("ProducerSettings: notifications API path must be non-empty string")
+	}
+	return nil
+}
+
+// DefaultProducerSettings are the default settings for the producer
+func DefaultProducerSettings() *ProducerSettings {
+	return &ProducerSettings{
+		MinPingPeriod:        time.Second,
+		ReconnectDelay:       3 * time.Second,
+		PongTimeout:          2 * time.Second,
+		PingPeriodPercentage: 60,
+		NotificationsAPIPath: "/v1/notifications",
+	}
+}
+
+// Message is the payload sent by the producer
+type Message struct {
+	// Notification is the notification that needs to be applied on the platform
+	Notification *types.Notification
+
+	// Resync indicates if the notifications stream was interrupted, so a full resync is needed
+	Resync bool
+}
+
+// NewProducer returns a configured producer for the given settings
+func NewProducer(producerSettings *ProducerSettings, smSettings *sm.Settings) (*Producer, error) {
+	notificationsURL, err := buildNotificationsURL(smSettings.URL, producerSettings.NotificationsAPIPath)
+	if err != nil {
+		return nil, err
+	}
+	return &Producer{
+		url:              notificationsURL,
+		producerSettings: *producerSettings,
+		smSettings:       *smSettings,
+	}, nil
+}
+
+func buildNotificationsURL(baseURL, notificationsPath string) (*url.URL, error) {
+	smURL, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, err
+	}
+	switch smURL.Scheme {
+	case "http":
+		smURL.Scheme = "ws"
+	case "https":
+		smURL.Scheme = "wss"
+	}
+	smURL.Path = path.Join(smURL.Path, notificationsPath)
+	return smURL, nil
+}
+
+// Start starts the producer in a new go-routine
+func (p *Producer) Start(ctx context.Context, messages chan *Message) {
+	go p.run(ctx, messages)
+}
+
+func (p *Producer) run(ctx context.Context, messages chan *Message) {
+	for {
+		needResync := p.lastNotificationRevision == 0
+		if err := p.connect(ctx); err != nil {
+			log.C(ctx).WithError(err).Error("could not connect websocket")
+			if err == errLastNotificationGone { // skip reconnect delay
+				p.lastNotificationRevision = 0
+				continue
+			}
+		} else {
+			if needResync {
+				messages <- &Message{Resync: true}
+			}
+			p.conn.SetPongHandler(func(string) error {
+				log.C(ctx).Debug("Received pong")
+				return p.conn.SetReadDeadline(time.Now().Add(p.readTimeout))
+			})
+
+			done := make(chan struct{}, 1)
+			childContext, stopChildren := context.WithCancel(ctx)
+			go p.readNotifications(childContext, messages, done)
```

Comment:
probably producer should produce all kinds of "Message"-s 

We could spawn a go.resync goroutine here with ticker that writes resync messages, and store the latestResyncTimeStamp in the producer. And the ticker writes a resync message if the currentmoment-lastresync > some configurable value otherwise it skips.

Reply:
The idea is that notifications.Producer is responsible only for notifications arriving on the websocket.
Regular resync should be implemented in another component.
We imagine code like this:
```go
messages := make(chan *notifications.Message, 1024)
p.notificationsProducer.Start(p.ctx, messages)
for {
  select {
    case m := <-messages:
      if m.Resync {
        // do resync - we cannot skip or delay this resync as there is a break in the notification flow (e.g. status 410), so a resync is mandatory
        // reset timer
      } else {
        // apply m.Notification
      }
    case <-timer.C:
       // do resync
       // reset timer
  }
}
```

Reply:
then you trigger resync in two different ways - either by writing a resync message or by ticker - my idea is to always trigger the full resync only if a message with resync is present. With your idea you have no way to guarantee the interval(enough time to pass between resyncs) between the ticker resyncs and the resyncs that would be triggered by resync messages

Reply:
discussed in person - ticker/timer will be in the producer. Will send messages when a tick occurs and will reset the timer when sending a resync message